### PR TITLE
druid-shell IME api, take two

### PR DESCRIPTION
**Pull request by cmyr**
This supersedes #1378; I'm going to use this branch to make some tweaks to the API, and eventually to integrate this into the druid textbox.

cc @lord, in case you'd like to follow along with some of this stuff.

The main change I'm envisioning currently is to have separate `acquire` and `release` methods for the locking, because druid needs some callback to know if it has to update the app data.

### CODE REVIEW

**Comment by cmyr** [comment]
@lord I've pushed a few more changes here that made sense in the process of doing #1636. Happy to have you take a look if you like, or else I'll self-review and do any further cleanup from there. :)

File: druid-shell/src/text.rs [hunk 1]
```diff
@@ -403,6 +403,7 @@ pub fn simulate_input<H: WinHandler + ?Sized>(
     event: KeyEvent,
 ) -> bool {
     if handler.key_down(event.clone()) {
+        eprintln!("key handled, returning early");
```
Comment (cmyr) [inline druid-shell/src/text.rs:406]: oops

**Comment by cmyr** [comment]
Okay I'm going to take a self-review on this, it's mostly the same code as #1378, which has already been reviewed.

**Comment by lord** [comment]
Sorry, just saw this! It looks great to me!

**Pull request merged**